metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala

```diff
@@ -1,902 +1,774 @@
- *  caller to shut down the client.
- * @param initializeParams
- *  Initialization parameters send by the client in the initialize request,
- *  which is the first request sent to the server by the client.
- */
+import org.eclipse.lsp4j.FoldingRange
+import org.eclipse.lsp4j.FoldingRangeRequestParams
+import org.eclipse.lsp4j.Hover
+import org.eclipse.lsp4j.Location
+import org.eclipse.lsp4j.ReferenceParams
+import org.eclipse.lsp4j.RenameFilesParams
+import org.eclipse.lsp4j.RenameParams
+import org.eclipse.lsp4j.SelectionRange
+import org.eclipse.lsp4j.SelectionRangeParams
+import org.eclipse.lsp4j.SemanticTokens
+import org.eclipse.lsp4j.SemanticTokensParams
+import org.eclipse.lsp4j.SignatureHelp
+import org.eclipse.lsp4j.SymbolInformation
+import org.eclipse.lsp4j.TextDocumentPositionParams
+import org.eclipse.lsp4j.TextEdit
+import org.eclipse.lsp4j.WorkspaceEdit
+import org.eclipse.lsp4j.WorkspaceSymbolParams
+import org.eclipse.lsp4j.jsonrpc.messages
+
 class MetalsLspService(
-    ec: ExecutionContextExecutorService,
-    sh: ScheduledExecutorService,
+    val folderServices: List[MetalsLspWorkspaceFolderService],
     serverInputs: MetalsServerInputs,
-    val workspace: AbsolutePath,
-    client: MetalsLanguageClient,
-    initializeParams: InitializeParams,
-) extends Cancelable
-    with ScalaLspService {
+    ec: ExecutionContext,
+    initializeParams: lsp4j.InitializeParams,
+    clientConfig: ClientConfiguration,
+) extends ScalaLspService {
   import serverInputs._
+  implicit val ex: ExecutionContext = ec
 
-  ThreadPools.discardRejectedRunnables("MetalsLanguageServer.sh", sh)
-  ThreadPools.discardRejectedRunnables("MetalsLanguageServer.ec", ec)
+  private val timerProvider: TimerProvider = new TimerProvider(time)
 
-  private val cancelables = new MutableCancelable()
-  val isCancelled = new AtomicBoolean(false)
+  def getServiceFor(uri: String): Option[MetalsLspWorkspaceFolderService] =
+    for { // where should we log info from here ???
+      path <- uri.toAbsolutePathSafe
+      workSpaceFolder <- folderServices
+        .filter(service => path.toNIO.startsWith(service.workspace.toNIO))
+        .sortBy(_.workspace.toNIO.getNameCount())
+        .lastOption
+    } yield workSpaceFolder
+
+  def getServiceForName(
+      workspaceFolder: String
+  ): Option[MetalsLspWorkspaceFolderService] =
+    for {
+      workSpaceFolder <- folderServices
+        .find(service => service.workspaceId == workspaceFolder)
+    } yield workSpaceFolder
 
-  override def cancel(): Unit = {
-    if (isCancelled.compareAndSet(false, true)) {
-      val buildShutdown = bspSession match {
-        case Some(session) => session.shutdown()
-        case None => Future.successful(())
-      }
-      try {
-        cancelables.cancel()
-      } catch {
-        case NonFatal(_) =>
-      }
-      try buildShutdown.asJava.get(100, TimeUnit.MILLISECONDS)
-      catch {
-        case _: TimeoutException =>
-      }
+  def execForWorkspaceFolder[R](
```

Comment:
```suggestion
  def inWorkspace[R](
```
probably enough

Reply:
It's actually called "workspace folder". Workspace is the whole thing. I'm afraid I'm not very consistent about that, but I'll fix it.